### PR TITLE
Feature/168263892 extract technology type from single cell experiments

### PR DESCRIPTION
**Pull request by lingyun1010**
Add an sdrf parser for single-cell technology type, which is actually the field `library construction`.

### CODE REVIEW

File: src/main/java/uk/ac/ebi/atlas/model/experiment/Experiment.java [hunk 1]
```diff
@@ -48,7 +50,8 @@
     private final boolean isPrivate;
     private String accessKey;
 
-    public Experiment(@NotNull ExperimentType type,
+    public Experiment(@Nullable Optional<List<String>> technologyType,
```
Comment (haideriqbal) [inline src/main/java/uk/ac/ebi/atlas/model/experiment/Experiment.java]: This can be changed to `@NotNull` as `Optional` can never be `null`. If it doesnt have a non-null value then it will be an empty `Optional`. More information here:
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Optional.html

File: src/main/java/uk/ac/ebi/atlas/model/experiment/differential/DifferentialExperiment.java [hunk 2]
```diff
@@ -29,7 +31,8 @@
 public class DifferentialExperiment extends Experiment<Contrast> {
     private final Set<Contrast> contrastsWithCttvPrimaryAnnotation;
 
-    public DifferentialExperiment(@NotNull ExperimentType experimentType,
+    public DifferentialExperiment(@Nullable Optional<List<String>> technologyType,
```
Comment (haideriqbal) [inline src/main/java/uk/ac/ebi/atlas/model/experiment/differential/DifferentialExperiment.java]: Same comment as above. Please review it at other places in code as well.

File: src/test/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParserIT.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        @AfterAll
+        void cleanDatabaseTables() {
+            ResourceDatabasePopulator populator = new ResourceDatabasePopulator();
+            populator.addScripts(new ClassPathResource("fixtures/experiment-delete.sql"));
+            populator.execute(dataSource);
+        }
+
+        @ParameterizedTest
+        @MethodSource("singleCellExperimentsProvider")
+        void testParserForSingleCell(String experimentAccession) {
```
Comment (haideriqbal) [inline src/test/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParserIT.java]: Please add `@DisplayName` to the test. It will help better understand what the test is supposed to do. Also a better test name maybe `testIfTechnologyTypePresentInSingleCellExperiment`

File: src/test/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParserOutputTest.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class SdrfParserOutputTest {
+
+    private SdrfParserOutput subject;
+
+    @Test
+    void testGetters() {
```
Comment (haideriqbal) [inline src/test/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParserOutputTest.java]: `@DisplayName` here as well please and test name can be `saneBasicMethods`

File: src/test/java/uk/ac/ebi/atlas/model/experiment/ExperimentBuilder.java [hunk 5]
```diff
@@ -50,6 +51,7 @@
     private final int dataProvidersSize = RNG.nextInt(3);
     private final int alternativeViewsSize = RNG.nextInt(4);
 
+    Optional<List<String>> technologyType = Optional.of(Arrays.asList(generateRandomExperimentAccession()));
```
Comment (haideriqbal) [inline src/test/java/uk/ac/ebi/atlas/model/experiment/ExperimentBuilder.java]: Is there any reason for assigning a list of `experimentAccessions` to `technologyType`?

File: src/main/java/uk/ac/ebi/atlas/model/experiment/differential/DifferentialExperiment.java (hunk 2 above)
Comment (lingyun1010) [reply src/main/java/uk/ac/ebi/atlas/model/experiment/differential/DifferentialExperiment.java]: Thanks. I corrected all the `NotNull` issues.

File: src/test/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParserIT.java (hunk 3 above)
Comment (lingyun1010) [reply src/test/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParserIT.java]: Add a display name as `parses technology type` and rename the test function as suggested.

File: src/test/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParserOutputTest.java (hunk 4 above)
Comment (lingyun1010) [reply src/test/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParserOutputTest.java]: Add a `DisplayName` as `gets a technology type list with unique values`.

File: src/test/java/uk/ac/ebi/atlas/model/experiment/ExperimentBuilder.java (hunk 5 above)
Comment (lingyun1010) [reply src/test/java/uk/ac/ebi/atlas/model/experiment/ExperimentBuilder.java]: No. I just would like to put a random string actually. I will reformat it.

File: src/main/java/uk/ac/ebi/atlas/utils/ExperimentInfo.java [hunk 6]
```diff
@@ -41,7 +41,7 @@ public void setExperimentAccession(String experimentAccession) {
     }
 
     public void setTechnologyType(Optional<List<String>> technologyType) {
-        this.technologyType = technologyType;
+        this.technologyType = technologyType.get();
```
Comment (haideriqbal) [inline src/main/java/uk/ac/ebi/atlas/utils/ExperimentInfo.java]: I will first check if value is present before assigning it using `Optional.isPresent` or `Optional.isEmpty`

**Comment by lingyun1010** [comment]
> Why do you need `SdrfParserOutput`? What advantages does it have over the bare return type? How is an empty list semantically different than an empty optional of a list?
> 
> Think about these questions and we can discuss before making further changes if you prefer.

File: src/main/java/uk/ac/ebi/atlas/model/experiment/Experiment.java [hunk 7]
```diff
@@ -48,7 +49,8 @@
     private final boolean isPrivate;
     private String accessKey;
 
-    public Experiment(@NotNull ExperimentType type,
+    public Experiment(@NotNull List<String> technologyType,
```
Comment (alfonsomunozpomer) [inline src/main/java/uk/ac/ebi/atlas/model/experiment/Experiment.java]: The constructor arguments in `Experiment` tend to express the importance of the passed values. Since `technologyType` is an attribute that’s going to be empty (at least for the time being) in bulk experiments and can take an indefinite set of values, move it after `species`. Also, the fully correct type is `@NotNull List<@NotNull String>`.

The same consideration applies to the method `buildExperimentInfo`.

File: src/main/java/uk/ac/ebi/atlas/model/experiment/Experiment.java [hunk 8]
```diff
@@ -28,6 +28,7 @@
 // The displayName is a bit confusing - it's used for baseline landing page and I think only there.
 // There's also a title which is fetched from the IDF file.
 public abstract class Experiment<R extends ReportsGeneExpression> implements Serializable {
+    private final List<String>  technologyType;
```
Comment (alfonsomunozpomer) [inline src/main/java/uk/ac/ebi/atlas/model/experiment/Experiment.java]: Use an immutable list and use `ImmutableList.copyOf` in the constructor like the rest of the collection-like arguments.

File: src/main/java/uk/ac/ebi/atlas/model/experiment/Experiment.java [hunk 9]
```diff
@@ -97,6 +100,8 @@ public Experiment(@NotNull ExperimentType type,
         this.accessKey = accessKey;
     }
 
+    public List<String>  getTechnologyType() { return technologyType; }
```
Comment (alfonsomunozpomer) [inline src/main/java/uk/ac/ebi/atlas/model/experiment/Experiment.java]: Change the return type (with `@NotNull`) after making this immutable.

File: src/main/java/uk/ac/ebi/atlas/model/experiment/ExperimentDesign.java [hunk 10]
```diff
@@ -48,6 +48,7 @@
     // Headers retrieved from the SDRF file, which maintain a curated order
     private Set<String> orderedSampleCharacteristicHeaders;
     private Set<String> orderedFactorHeaders;
+    private Set<String> orderedCommentHeaders;
```
Comment (alfonsomunozpomer) [inline src/main/java/uk/ac/ebi/atlas/model/experiment/ExperimentDesign.java]: Why is this needed? Who’s using it?

File: src/main/java/uk/ac/ebi/atlas/testutils/MockExperiment.java [hunk 11]
```diff
@@ -19,6 +19,7 @@
 import java.util.Arrays;
 import java.util.Date;
 import java.util.List;
+import java.util.Optional;
```
Comment (alfonsomunozpomer) [inline src/main/java/uk/ac/ebi/atlas/testutils/MockExperiment.java]: Stale import.

File: src/main/java/uk/ac/ebi/atlas/testutils/MockExperiment.java [hunk 12]
```diff
@@ -148,6 +149,7 @@ public static BaselineExperiment createBaselineExperiment(String accession,
                                                               List<String> pubmedIds,
                                                               List<String> dois) {
         return new BaselineExperiment(
+                Arrays.asList("technologyType"),
```
Comment (alfonsomunozpomer) [inline src/main/java/uk/ac/ebi/atlas/testutils/MockExperiment.java]: Use `ImmutableList.of`.

File: src/main/java/uk/ac/ebi/atlas/utils/ExperimentInfo.java [hunk 13]
```diff
@@ -5,10 +5,12 @@
 
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Optional;
```
Comment (alfonsomunozpomer) [inline src/main/java/uk/ac/ebi/atlas/utils/ExperimentInfo.java]: Stale import.

File: src/test/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParserIT.java [hunk 14]
```diff
@@ -0,0 +1,70 @@
+package uk.ac.ebi.atlas.experimentimport.sdrf;
```
Comment (alfonsomunozpomer) [inline src/test/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParserIT.java]: You can var-ify most types in this test.

File: src/test/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParserIT.java [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+            populator.execute(dataSource);
+        }
+
+        @ParameterizedTest
+        @MethodSource("singleCellExperimentsProvider")
+        @DisplayName("parses technology type")
+        void testIfTechnologyTypePresentInSingleCellExperiment(String experimentAccession) {
+            SdrfParser sdrfParser = new SdrfParser(
+                    new DataFileHub(dataFilesPath.resolve("scxa")));
+            List<String> result = sdrfParser.parse(experimentAccession);
```
Comment (alfonsomunozpomer) [inline src/test/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParserIT.java]: You don’t need this throw-away variable.

File: src/main/java/uk/ac/ebi/atlas/trader/factory/BaselineExperimentFactory.java [hunk 16]
```diff
@@ -29,7 +30,8 @@ public BaselineExperimentFactory(ConfigurationTrader configurationTrader,
     @Override
     public BaselineExperiment create(ExperimentDto experimentDto,
                                      ExperimentDesign experimentDesign,
-                                     IdfParserOutput idfParserOutput) {
+                                     IdfParserOutput idfParserOutput,
+                                     SdrfParser sdrfParser) {
```
Comment (alfonsomunozpomer) [inline src/main/java/uk/ac/ebi/atlas/trader/factory/BaselineExperimentFactory.java]: I think that here, and **in all factories**, you want the technology type rather than the parser. Currently there’s a compile error in `ScxaExperimentRepository` because of this.

File: src/main/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParser.java [hunk 17]
```diff
@@ -17,14 +18,28 @@
...
     private final DataFileHub dataFileHub;
 
     @Inject
     public SdrfParser(DataFileHub dataFileHub) {
         this.dataFileHub = dataFileHub;
     }
 
+    public List<String> parse(String experimentAccession) {
+        try (TsvStreamer sdrfStreamer = dataFileHub.getExperimentFiles(experimentAccession).sdrf.get()) {
+            var streamer = sdrfStreamer.get().collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (lingyun1010) [reply src/main/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParser.java]: I rename the `streamer` as `lines` and return as an `ImmutableList`.

File: src/main/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParser.java [hunk 18]
```diff
@@ -17,14 +18,28 @@
...
+            var technologyTypeColumnIndex = Arrays.asList(streamer.stream().findFirst().get())
+                    .indexOf(TECHNOLOGY_TYPE_ID);
+
+            return streamer.stream()
+                            .skip(1)
+                            .map(line -> Arrays.asList(line).get(technologyTypeColumnIndex))
+                            .distinct()
+                            .collect(Collectors.toList()
```
[the remark this answers is not in the document]
Comment (lingyun1010) [reply src/main/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParser.java]: I changed the return type as `ImmutableList`.

File: src/main/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParser.java [hunk 19]
```diff
@@ -17,14 +18,28 @@
...
 
+    public List<String> parse(String experimentAccession) {
+        try (TsvStreamer sdrfStreamer = dataFileHub.getExperimentFiles(experimentAccession).sdrf.get()) {
+            var streamer = sdrfStreamer.get().collect(Collectors.toList());
+            var technologyTypeColumnIndex = Arrays.asList(streamer.stream().findFirst().get())
+                    .indexOf(TECHNOLOGY_TYPE_ID);
+
+            return streamer.stream()
+                            .skip(1)
+                            .map(line -> Arrays.asList(line).get(technologyTypeColumnIndex))
```
[the remark this answers is not in the document]
Comment (lingyun1010) [reply src/main/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParser.java]: Done.

File: src/main/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParser.java [hunk 20]
```diff
@@ -17,14 +18,28 @@
...
 
     @Inject
     public SdrfParser(DataFileHub dataFileHub) {
         this.dataFileHub = dataFileHub;
     }
 
+    public List<String> parse(String experimentAccession) {
+        try (TsvStreamer sdrfStreamer = dataFileHub.getExperimentFiles(experimentAccession).sdrf.get()) {
+            var streamer = sdrfStreamer.get().collect(Collectors.toList());
+            var technologyTypeColumnIndex = Arrays.asList(streamer.stream().findFirst().get())
```
[the remark this answers is not in the document]
Comment (lingyun1010) [reply src/main/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParser.java]: Thanks for the comments. I refactor codes as suggested.

File: src/main/java/uk/ac/ebi/atlas/model/experiment/Experiment.java (hunk 7 above)
Comment (lingyun1010) [reply src/main/java/uk/ac/ebi/atlas/model/experiment/Experiment.java]: Refactored as `List<@NotNull String> technologyType`.

File: src/main/java/uk/ac/ebi/atlas/model/experiment/Experiment.java (hunk 8 above)
Comment (lingyun1010) [reply src/main/java/uk/ac/ebi/atlas/model/experiment/Experiment.java]: Refactored as `this.technologyType = ImmutableList.copyOf(technologyType);`

File: src/main/java/uk/ac/ebi/atlas/model/experiment/Experiment.java (hunk 9 above)
Comment (lingyun1010) [reply src/main/java/uk/ac/ebi/atlas/model/experiment/Experiment.java]: Done.

File: src/main/java/uk/ac/ebi/atlas/model/experiment/ExperimentDesign.java (hunk 10 above)
Comment (lingyun1010) [reply src/main/java/uk/ac/ebi/atlas/model/experiment/ExperimentDesign.java]: Sorry this is a stale junk code I tested before.  I will delete it.

File: src/main/java/uk/ac/ebi/atlas/testutils/MockExperiment.java (hunk 11 above)
Comment (lingyun1010) [reply src/main/java/uk/ac/ebi/atlas/testutils/MockExperiment.java]: Deleted.

File: src/main/java/uk/ac/ebi/atlas/testutils/MockExperiment.java (hunk 12 above)
Comment (lingyun1010) [reply src/main/java/uk/ac/ebi/atlas/testutils/MockExperiment.java]: Done.

File: src/main/java/uk/ac/ebi/atlas/utils/ExperimentInfo.java (hunk 13 above)
Comment (lingyun1010) [reply src/main/java/uk/ac/ebi/atlas/utils/ExperimentInfo.java]: Removed.

File: src/test/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParserIT.java (hunk 15 above)
Comment (lingyun1010) [reply src/test/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParserIT.java]: Reformat it without defining a variable.

File: src/main/java/uk/ac/ebi/atlas/trader/factory/BaselineExperimentFactory.java (hunk 16 above)
Comment (lingyun1010) [reply src/main/java/uk/ac/ebi/atlas/trader/factory/BaselineExperimentFactory.java]: Pass technology type string instead of `SdrfParser`.

File: src/main/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParser.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (alfonsomunozpomer) [reply src/main/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParser.java]: Yep, that was the name I had in mind. :wink:

File: src/main/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParser.java [hunk 21]
```diff
@@ -9,22 +10,39 @@
...
+            var lines = sdrfStreamer.get().collect(toImmutableList());
+            var technologyTypeColumnIndex = ImmutableList.copyOf(lines.get(0)).indexOf(TECHNOLOGY_TYPE_ID);
+
+            return technologyTypeColumnIndex > 0 ?
+                    lines.stream()
+                            .skip(1)
+                            .map(line -> ImmutableList.copyOf(line).get(technologyTypeColumnIndex))
+                            .distinct()
+                    .collect(toImmutableList()) :
+                    ImmutableList.of("");
```
Comment (alfonsomunozpomer) [inline src/main/java/uk/ac/ebi/atlas/experimentimport/sdrf/SdrfParser.java]: I think you should return an empty list, i.e. `ImmutableList.of()` if the SDRF doesn’t have the *library construction* header.
